public/components/custom_panels/helpers/utils.tsx

```diff
@@ -311,44 +329,107 @@ const createCatalogVisualizationMetaData = (
   };
 };
 
+const updateCatalogVisualizationQuery = ({
+  catalogSourceName,
+  catalogTableName,
+  aggregation,
+  attributesGroupBy,
+  startTime,
+  endTime,
+  spanParam,
+}: {
+  catalogSourceName: string;
+  catalogTableName: string;
+  aggregation: string;
+  attributesGroupBy: string[];
+  startTime: string;
+  endTime: string;
+  spanParam: string | undefined;
+}) => {
+  // source=my_prometheus.query_range('avg by(attribue1, attribuyte2) (prometheus_requests_total)', 1686694425, 1686700130, 14)
+  const attributesGroupString = attributesGroupBy.toString();
+  const startEpochTime = convertDateTimeToEpoch(startTime);
+  const endEpochTime = convertDateTimeToEpoch(endTime, false);
+  const promQuery =
+    attributesGroupBy.length === 0
+      ? catalogTableName
+      : `${aggregation} by(${attributesGroupString}) (${catalogTableName})`;
+
+  return `source = ${catalogSourceName}.query_range('${promQuery}', ${startEpochTime}, ${endEpochTime}, '${spanParam}')`;
+};
+
 // Creates a catalogVisualization for a runtime catalog based PPL query and runs getQueryResponse
-export const renderCatalogVisualization = async (
-  http: CoreStart['http'],
-  pplService: PPLService,
-  catalogSource: string,
-  startTime: string,
-  endTime: string,
-  filterQuery: string,
-  spanParam: string | undefined,
-  setVisualizationTitle: React.Dispatch<React.SetStateAction<string>>,
-  setVisualizationType: React.Dispatch<React.SetStateAction<string>>,
-  setVisualizationData: React.Dispatch<React.SetStateAction<Plotly.Data[]>>,
-  setVisualizationMetaData: React.Dispatch<React.SetStateAction<undefined>>,
-  setIsLoading: React.Dispatch<React.SetStateAction<boolean>>,
-  setIsError: React.Dispatch<React.SetStateAction<VizContainerError>>,
-  spanResolution?: string
-) => {
+export const renderCatalogVisualization = async ({
+  http,
+  pplService,
+  catalogSource,
+  startTime,
+  endTime,
+  filterQuery,
+  spanParam,
+  setVisualizationTitle,
+  setVisualizationType,
+  setVisualizationData,
+  setVisualizationMetaData,
+  setIsLoading,
+  setIsError,
+  spanResolution,
+  queryMetaData,
+}: {
+  http: CoreStart['http'];
+  pplService: PPLService;
+  catalogSource: string;
+  startTime: string;
+  endTime: string;
+  filterQuery: string;
+  spanParam: string | undefined;
+  setVisualizationTitle: React.Dispatch<React.SetStateAction<string>>;
+  setVisualizationType: React.Dispatch<React.SetStateAction<string>>;
+  setVisualizationData: React.Dispatch<React.SetStateAction<Plotly.Data[]>>;
+  setVisualizationMetaData: React.Dispatch<React.SetStateAction<undefined>>;
+  setIsLoading: React.Dispatch<React.SetStateAction<boolean>>;
+  setIsError: React.Dispatch<React.SetStateAction<VizContainerError>>;
+  spanResolution?: string;
+  queryMetaData?: MetricType;
```

Comment:
Question: Just curious does converting params to object help in any specific way?

Reply:
Over 3 params makes it very hard to organize or arrange the parameter list in the fn. definition or in calls.
Over the course of investigation on this PR, added/removed several parameters in the middle of the call.

Reply:
I actually thinking if this is the indicator of having a function that is dealing too many tasks. Even switching to use an object, I don't see they are all logically connected. Maybe split params to different objects or split this big function into task based smaller ones. Just put some thoughts here.

Reply:
I think I agree.  There are other wider-ranging refactors that I'll look into in a different PR